sycl/source/detail/scheduler/commands.cpp

```diff
@@ -1797,71 +1797,87 @@ cl_int ExecCGCommand::enqueueImp() {
+          break;
+        }
         }
-        break;
-      }
-      case kernel_param_kind_t::kind_std_layout: {
-        Plugin.call<PiApiKind::piKernelSetArg>(Kernel, Arg.MIndex, Arg.MSize,
-                                               Arg.MPtr);
-        break;
-      }
-      case kernel_param_kind_t::kind_sampler: {
-        sampler *SamplerPtr = (sampler *)Arg.MPtr;
-        RT::PiSampler Sampler =
-            detail::getSyclObjImpl(*SamplerPtr)->getOrCreateSampler(Context);
-        Plugin.call<PiApiKind::piKernelSetArg>(Kernel, Arg.MIndex,
-                                               sizeof(cl_sampler), &Sampler);
-        break;
-      }
-      case kernel_param_kind_t::kind_pointer: {
-        Plugin.call<PiApiKind::piextKernelSetArgPointer>(Kernel, Arg.MIndex,
-                                                         Arg.MSize, Arg.MPtr);
-        break;
-      }
       }
-    }
 
-    adjustNDRangePerKernel(NDRDesc, Kernel,
-                           *(detail::getSyclObjImpl(MQueue->get_device())));
+      adjustNDRangePerKernel(NDRDesc, Kernel,
+                             *(detail::getSyclObjImpl(MQueue->get_device())));
 
-    // Some PI Plugins (like OpenCL) require this call to enable USM
-    // For others, PI will turn this into a NOP.
-    Plugin.call<PiApiKind::piKernelSetExecInfo>(Kernel, PI_USM_INDIRECT_ACCESS,
-                                                sizeof(pi_bool), &PI_TRUE);
+      // Some PI Plugins (like OpenCL) require this call to enable USM
+      // For others, PI will turn this into a NOP.
+      Plugin.call<PiApiKind::piKernelSetExecInfo>(
+          Kernel, PI_USM_INDIRECT_ACCESS, sizeof(pi_bool), &PI_TRUE);
 
-    // Remember this information before the range dimensions are reversed
-    const bool HasLocalSize = (NDRDesc.LocalSize[0] != 0);
+      // Remember this information before the range dimensions are reversed
+      const bool HasLocalSize = (NDRDesc.LocalSize[0] != 0);
 
-    ReverseRangeDimensionsForKernel(NDRDesc);
+      ReverseRangeDimensionsForKernel(NDRDesc);
+      pi_result Error = Plugin.call_nocheck<PiApiKind::piEnqueueKernelLaunch>(
+          MQueue->getHandleRef(), Kernel, NDRDesc.Dims,
+          &NDRDesc.GlobalOffset[0], &NDRDesc.GlobalSize[0],
+          HasLocalSize ? &NDRDesc.LocalSize[0] : nullptr, RawEvents.size(),
+          RawEvents.empty() ? nullptr : &RawEvents[0], &Event);
+      return Error;
+    };
 
-    pi_result Error = Plugin.call_nocheck<PiApiKind::piEnqueueKernelLaunch>(
-        MQueue->getHandleRef(), Kernel, NDRDesc.Dims, &NDRDesc.GlobalOffset[0],
-        &NDRDesc.GlobalSize[0], HasLocalSize ? &NDRDesc.LocalSize[0] : nullptr,
-        RawEvents.size(), RawEvents.empty() ? nullptr : &RawEvents[0], &Event);
+    pi_result Error = PI_SUCCESS;
+    if (KernelMutex != nullptr) {
+      std::lock_guard<std::mutex> Lock(*KernelMutex);
+      Error = SetKernelParamsAndLaunch();
+    } else {
+      Error = SetKernelParamsAndLaunch();
```

Comment:
As far as I understand, we still will get a race-condition if we submit the same kernel with `nullptr != ExecKernel->MSyclKernel` in multiple threads, isn't it?
Is this case covered by some tests?

Reply:
Right, I covered this case: `nullptr != ExecKernel->MSyclKernel`.
No, at the moment, but I will add the test in our test suite nearby the previous regression test.
So in the scope of the PR, I have done all.